src/storage/src/hummock/local_version_manager.rs

```diff
@@ -552,27 +607,135 @@ impl LocalVersionManager {
+                            .buffer_tracker
+                            .global_upload_task_size
+                            .fetch_add(size, Relaxed);
+                        epoch_join_handle
+                            .entry(epoch)
+                            .or_insert_with(Vec::new)
+                            .push(join_handle);
+                    } else {
+                        break;
+                    }
+                }
+            };
+
+        let grant_write_request = |size, sender: oneshot::Sender<()>| {
+            local_version_manager
+                .buffer_tracker
+                .global_buffer_size
+                .fetch_add(size, Relaxed);
+            let _ = sender.send(()).inspect_err(|err| {
+                error!("unable to send write request response: {:?}", err);
+            });
+        };
+
+        let mut pending_write_requests: VecDeque<(usize, oneshot::Sender<()>)> = VecDeque::new();
+
+        // While the current Arc is not the only strong reference to the local version manager
+        while Arc::strong_count(&local_version_manager) > 1 {
+            if let Some(event) = buffer_size_change_receiver.recv().await {
+                match event {
+                    SharedBufferEvent::WriteRequest(size, sender) => {
+                        if local_version_manager.buffer_tracker.can_write() {
+                            grant_write_request(size, sender);
+                        } else {
+                            pending_write_requests.push_back((size, sender));
+                        }
+                    }
+                    SharedBufferEvent::BatchAdded => {
+                        // Only check and flush shared buffer after batch has been added to shared
+                        // buffer.
+                        try_flush_shared_buffer(&syncing_epoch, &mut epoch_join_handle);
+                    }
+                    SharedBufferEvent::BufferRelease(size) => {
+                        local_version_manager
+                            .buffer_tracker
+                            .global_buffer_size
+                            .fetch_sub(size, Relaxed);
+                        while !pending_write_requests.is_empty()
+                            && local_version_manager.buffer_tracker.can_write()
+                        {
+                            let (size, sender) = pending_write_requests.pop_front().unwrap();
+                            grant_write_request(size, sender);
+                        }
+                    }
+                    SharedBufferEvent::UploadTaskFinish(size) => {
+                        local_version_manager
+                            .buffer_tracker
+                            .global_upload_task_size
+                            .fetch_sub(size, Relaxed);
+                        // An upload task has finished. There may be possibility that we need a new
+                        // upload task.
```

Comment:
Can you explain more on why we need a new upload task without waiting for buffer release? Writes only look at buffer size, not buffer size - upload task size.